### PR TITLE
[SYCL][NVPTX] Obey -fcuda-short-ptr when compiling SYCL for NVPTX

### PR DESCRIPTION
**Pull request by frasercrmck**
This flag turns pointers to CUDA's `shared`, `const`, and `local` address spaces into 32-bit pointers. This can potentially save on registers used for addressing calculations.

This option was being accepted by the frontend when compiling SYCL code, but was then reporting an error that the backend datalayout doesn't match the expected target description. This was because the option wasn't being caught by all parts of the toolchain, leading to inconsistencies.

This PR allows users to pass the option if they wish. They will see a warning that the compiler is linking against a libclc/libspirv that hasn't been compiled with this option, but this is likely harmless since libspirv doesn't manipulate pointers.

### CODE REVIEW

**Comment by hdelan** [comment]
https://www.youtube.com/watch?v=koJlIGDImiU

**Review by konradkusiak97 — APPROVED** [review]
This is great! And it looks good to me but maybe other people from @intel/llvm-reviewers-cuda want to also double-check.

File: clang/lib/Driver/ToolChains/Cuda.cpp [hunk 1]
```diff
@@ -1010,6 +1010,11 @@ void CudaToolChain::addClangTargetOptions(
     CC1Args.push_back(DriverArgs.MakeArgString(LibSpirvFile));
   }
 
+  bool UseShortPtr = DeviceOffloadingKind == Action::OFK_SYCL;
+  if (DriverArgs.hasFlag(options::OPT_fcuda_short_ptr,
+                         options::OPT_fno_cuda_short_ptr, UseShortPtr))
+    CC1Args.append({"-mllvm", "--nvptx-short-ptr"});
+
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Cuda.cpp:1017]: Driver changes look OK - please add associated LIT tests that reflect the new default settings when using `-fsycl`

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -8537,9 +8537,10 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     }
   }
 
-  if (IsCuda) {
+  if (IsCuda || (IsSYCLDevice && Triple.isNVPTX())) {
```
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/Clang.cpp:8540]: nit: Please add some comments for the newly added code.

File: libclc/CMakeLists.txt [hunk 3]
```diff
@@ -450,10 +450,15 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
       list(APPEND flags -D__unix__)
     endif()
 
+    set(spirv_flags ${flags})
+    if( ARCH STREQUAL nvptx OR ARCH STREQUAL nvptx64 )
+      list(APPEND spirv_flags -Xclang -fcuda-short-ptr -mllvm -nvptx-short-ptr)
```
Comment (Naghasan) [inline libclc/CMakeLists.txt:455]: I *think* this is ok, but you could end up linking the libspirv (with short ptr) to a program compiled without. As builtins don't write pointers, I don't think this is an issue but would be good to test prior to merging and document this.

File: clang/lib/Driver/ToolChains/Cuda.cpp (hunk 1 above)
Comment (frasercrmck) [reply clang/lib/Driver/ToolChains/Cuda.cpp]: Done, thanks. Let me know if they're acceptable.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (frasercrmck) [reply clang/lib/Driver/ToolChains/Clang.cpp:8547]: Done, thanks

File: libclc/CMakeLists.txt (hunk 3 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt]: Note I've now reduced the scope of this PR and the option is no longer enabled by default. Thus libclc/libspirv is no longer compiled with this option. If a user passes `-fcuda-short-ptr` they'll see a warning while linking libclc/libspirv which I think is correct to do.

But yes in general we need to consider whether this is okay. I also think it's okay, and I've run several benchmarks with it and not seen a problem. Ideally we'd compile two versions of libspirv for NVPTX, but I don't know if it's going to be worth it for this relatively obscure option.

File: clang/test/Driver/sycl-nvptx-short-ptr.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// REQUIRES: nvptx-registered-target
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-nvptx-short-ptr.cpp:1]: This shouldn't be needed - the test is just exercising driver behaviors, so no specific environment target should be required.

Comment (frasercrmck) [reply clang/test/Driver/sycl-nvptx-short-ptr.cpp]: I was wondering about that - thanks. I just copied it from another test. But as you say, all clang targets should always be available.

Anyway, fixed

Comment (bader) [reply clang/test/Driver/sycl-nvptx-short-ptr.cpp]: @frasercrmck, just to be clear. What Mike says is that this test doesn't use nvptx target, so it shouldn't require it. The test verifies driver's behavior only.

> all clang targets should always be available

This is not true. We build NVPTX target in this repository's CI environment, but some developers might not build this target. That's why `nvptx-registered-target` feature exists in general, but it's not relevant for this specific case.

Comment (frasercrmck) [reply clang/test/Driver/sycl-nvptx-short-ptr.cpp]: Right yes, sorry, that's what I meant: that no clang code is conditionally compiled depending on which LLVM targets are available. I wasn't saying that each LLVM target is always available. Unfortunately even upstream LLVM get confused about the purpose of this feature (e.g., https://github.com/llvm/llvm-project/pull/94055).

File: clang/test/CodeGenSYCL/nvptx-short-ptr.cpp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -fsycl-is-device -disable-llvm-passes \
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/nvptx-short-ptr.cpp:1]: Can you add a comment describing what you're checking for in the layout here?

Comment (frasercrmck) [reply clang/test/CodeGenSYCL/nvptx-short-ptr.cpp:8]: done, thanks

**Review by mdtoguchi — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Comment by frasercrmck** [comment]
@intel/llvm-gatekeepers this is ready to merge, thank you

**Pull request merged**